torch/functional.py

```diff
@@ -1291,5 +1293,112 @@ def _lu_no_infos(A, pivot=True, get_infos=False, out=None):
     func_name='lu')
 lu.__doc__ = _lu_impl.__doc__
 
+def clamp(input, min=None, max=None, out=None):
+    r"""
+    Clamp all elements in :attr:`input` into the range `[` :attr:`min`, :attr:`max` `]` and return
+    a resulting tensor:
+
+    .. math::
+        y_i = \min(\text{max}_i, \min(\text{min}_i, x_i))
+
+    The :attr:`input` and arguments :attr:`min` and attr:`max` should have matching types.
```

Comment:
Is this actually true for your new functions? It looks like they support mismatched input dtypes.

Reply:
It was close, but note quite there yet. I've now made changes such that you can use mixed types and hence removed this statement.

But there is one detail - the output dtype of `clamp` and `clamp_with_tensors` can be different if the min/max types are not the same as input. For `clamp` the input dtype is preserved, and for `clamp_with_tensors` the output uses standard type promotion from all its arguments to decide the output type. I've added a statement about this in the docstring.

If you'd like to also use type promotion for `clamp` with scalar args, then I'd have to wrap both scalar arguments in `wrapped_scalar_tensor` (this seems to be the norm, for example in math operations with scalars, e.g. `add`), which I've noticed can result in significant performance decrease. And since `clamp` is a quite widely used operator (e.g. ReLU layers),  I thought it would be best to keep the original behaviour/performance in this case.